src/shared/userdb.c

```diff
@@ -1205,6 +1369,47 @@ int membershipdb_iterator_get(
                 iterator->found_group_name = mfree(iterator->found_group_name);
         }
 
+        for (; iterator->dropins && iterator->dropins[iterator->current_dropin]; iterator->current_dropin++) {
+                const char *i = iterator->dropins[iterator->current_dropin], *e, *c;
+                _cleanup_free_ char *un = NULL, *gn = NULL;
+
+                e = endswith(i, ".membership");
+                if (!e)
+                        continue;
+
+                c = memchr(i, ':', e - i);
```

Comment:
Could extract_first_word&friends be used here instead?

Reply:
it could. but it doesn't get us much in this case, as nothing of the niceties it offers are needed/desired here. i.e. there's no escaping here, no quoting, no separator coalescing. It's so simple that memchr() i think is nicer.

I mean, we could rewrite this, but it's a bit less sexy, I think, as we'd first have to chop off the ".membership" suffix before we can call extract_first_word() on it, but this means doing the strdup() earlier, and in most loop terations. it's kinda nice if we can do superficial filtering out early, and only then strdup and do the in-depth verification. because that means that in almost all cases the strdup() would be worth it, since the superficial filtering will often filter stuff out, the in-depth verification is mostly just a safety net, that protects us from misuse but should never be hit in "clean" cases. In this case, the superficial check to me is whether the fname ends in ".membership" and that a ":" is included in the fname. The in-depth check if the split out user/group name pass valid_user_group_name().

I mean the idea is that we use the same drop-in dir for *.group, *.user, *.membership. That means that we'll usually have a bunch of files there that do not match what we are looking for, hence I think it's worth filtering thta stuff out early, before doing strdup().

I guess one could say: both approaches are absolutely fine, it's all a bit vague and not clear cut, I'd say.

Does that make any sense?

Reply:
yeah that makes sense